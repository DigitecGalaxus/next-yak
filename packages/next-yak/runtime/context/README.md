# Yak Theme Context

The yak theme context allows users to configure the theme of their yak application.

It works for react-server components and for normal react components.

## Usage

The context will be available in every `styled` and `css` literal:

```tsx
const Headline = styled.h1`
    color: ${props => props.theme.colors.primary};
`;
```


## React Server Components

To use the yak theme context in a react server component, you need to place a `yak.context.tsx` in
the same directory as your Next.js config.

```tsx
// yak.context.tsx
export function getYakThemeContext() {
    return {
        colors: {
            primary: "red",
            secondary: "blue",
        },
    }
}

declare module "next-yak/context" {
    export interface DefaultTheme extends ReturnType<typeof getYakThemeContext> { }
}
```

In your root layout component you have to forward the entire response of `getYakThemeContext`
context to the client:

```tsx
// layout.tsx
import { YakThemeProvider } from "next-yak";

export function Layout({ children }) {
    return (
        <YakThemeProvider>
            {children}
        </YakThemeProvider>
    );
}
```

Internally this works because of an alias set by `withYak` inside the NextJs config.

## React Components

If you are using server components you can **skip this section**.  

To use the yak theme context in a normal react component, you can just use the `YakThemeProvider`:

```tsx
// _app.tsx
import { YakThemeProvider } from "next-yak/context";

const theme = {
    colors: {
        primary: "red",
        secondary: "blue",
    },
};

export function App({ children }) {
    return (
        <YakThemeProvider value={theme}>
            {children}
        </YakThemeProvider>
    );
}
```