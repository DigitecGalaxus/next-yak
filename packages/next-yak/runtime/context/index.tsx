"use client";
//
// This file is the client component (browser & ssr) version of index.server.tsx
//
import React, { ReactNode, createContext, useContext } from "react";

export interface YakTheme {}

/**
 * The yak theme context
 * @see https://github.com/jantimon/next-yak/blob/main/packages/next-yak/runtime/context/README.md
 */
const YakContext = createContext<YakTheme>({});

/**
 * Returns the current yak theme context
 *
 * @see https://github.com/jantimon/next-yak/blob/main/packages/next-yak/runtime/context/README.md
 */
export const useTheme = (): YakTheme => useContext(YakContext);

/**
 * Yak theme context provider
 *
 * @see https://github.com/jantimon/next-yak/blob/main/packages/next-yak/runtime/context/README.md
 */
export const YakThemeProvider = ({
  children,
  theme = {},
}: {
  children: ReactNode;
  theme?: YakTheme;
}) => <YakContext.Provider value={theme}>{children}</YakContext.Provider>;
