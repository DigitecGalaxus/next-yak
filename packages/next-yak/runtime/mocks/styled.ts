import { styled as StyledFactory } from "../styled.js"
export const styled = new Proxy(
    StyledFactory as any,
    {
      get(target, TagName: keyof JSX.IntrinsicElements) {
        return target(TagName);
      },
    },
  );