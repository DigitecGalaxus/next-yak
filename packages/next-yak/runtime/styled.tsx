import { FunctionComponent } from "react";
import { CSSInterpolation, css } from "./cssLiteral";
import React from "react";

//
// The `styled()` and `styled.` API
//
// The API design is inspired by styled-components:
// https://github.com/styled-components/styled-components/blob/main/packages/styled-components/src/constructors/styled.tsx
// https://github.com/styled-components/styled-components/blob/main/packages/styled-components/src/models/StyledComponent.ts
//

type HtmlTags = keyof JSX.IntrinsicElements;

type Merge<T, U> = Omit<T, keyof U>;

type YakAttributes<T> = <TNew extends Record<string, unknown>>(
  attrArgs: ((props: T & TNew) => TNew & T) | (TNew & T)
) => YakTemplateString<Merge<T, TNew>>;

type YakTemplateString<T> = <TCSSProps extends Record<string, unknown>>(
  styles: TemplateStringsArray,
  ...values: Array<CSSInterpolation<TCSSProps>>
) => FunctionComponent<TCSSProps & T>;

type YakWithAttributes<T> = {
  <TCSSProps extends Record<string, unknown>>(
    styles: TemplateStringsArray,
    ...values: Array<CSSInterpolation<TCSSProps>>
  ): FunctionComponent<TCSSProps & T>;
  attrs: YakAttributes<T>;
};

type YakLiteralComponents = {
  [Tag in HtmlTags]: YakWithAttributes<JSX.IntrinsicElements[Tag]>;
};

type YakStyledComponentFunction = <T extends {}>(
  component: FunctionComponent<T>
) => YakWithAttributes<T>;

type YakStyled = YakStyledComponentFunction & YakLiteralComponents;

function StyledFactory<T>(
  Component: FunctionComponent<T>
): YakWithAttributes<T>;
function StyledFactory<T extends HtmlTags>(
  Component: T
): YakWithAttributes<JSX.IntrinsicElements[T]>;
function StyledFactory<T>(Component: HtmlTags | FunctionComponent<T>) {
  return Object.assign<YakTemplateString<T>, { attrs: YakAttributes<T> }>(
    (styles, ...values) => {
      return (props) => {
        const runtimeStyles = css(styles, ...values)(props as any);
        const filteredProps =
          typeof Component === "string"
            ? removePrefixedProperties(props)
            : props;
        return (
          <Component
            {...filteredProps}
            style={{ ...(props.style || {}), ...runtimeStyles.style }}
            className={
              (props.className ? props.className + " " : "") +
              runtimeStyles.className
            }
          />
        );
      };
    },
    {
      attrs: (attrsProps) => {
        return (styles, ...values) => {
          return (_props) => {
            const newProps =
              typeof attrsProps === "function"
                ? //@ts-expect-error
                  attrsProps(_props)
                : attrsProps;
            const props = {
              ..._props,
              ...newProps,
              className: mergeClassNames(
                _props.className as string,
                newProps.className as string
              ),
              style: { ...(_props.style || {}), ...(newProps.style || {}) },
              // children: _props.children,
              // className: _props.className,
              // style: _props.style,
            };
            const runtimeStyles = css(styles, ...values)(props as any);
            const filteredProps =
              typeof Component === "string"
                ? removePrefixedProperties(props)
                : props;
            return (
              <Component
                {...filteredProps}
                style={{ ...(props.style || {}), ...runtimeStyles.style }}
                className={mergeClassNames(
                  props.className as string,
                  runtimeStyles.className as string
                )}
              />
            );
          };
        };
      },
    }
  );
}

/**
 * The `styled` method works perfectly on all of your own or any third-party component,
 * as long as they attach the passed className prop to a DOM element.
 *
 * @usage
 *
 * ```tsx
 * const StyledLink = styled(Link)`
 *  color: #BF4F74;
 *  font-weight: bold;
 * `;
 * ```
 */
export const styled = new Proxy(StyledFactory, {
  get(target, TagName) {
    if (typeof TagName !== "string") {
      throw new Error("Only string tags are supported");
    }
    return target(TagName as keyof JSX.IntrinsicElements);
  },
}) as YakStyled;

// Remove all entries that start with a $ sign
function removePrefixedProperties<T extends Record<string, unknown>>(obj: T) {
  const result = {} as T;
  for (const key in obj) {
    if (!key.startsWith("$")) {
      result[key] = obj[key];
    }
  }
  return result;
}

const mergeClassNames = (a?: string, b?: string) => {
  if (!a) return b;
  if (!b) return a;
  return a + " " + b;
};
