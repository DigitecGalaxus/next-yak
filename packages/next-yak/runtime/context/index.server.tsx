//
// This file is the react-server component version of index.tsx
//

// @ts-ignore - in the current @types/react "cache" is not typed
import React, { ReactNode, cache } from "react";
import { YakThemeProvider as YakThemeClientProvider } from "./index.js";

// the following import might be changed by
// the user config in withYak to point to their own
// context
import { getYakThemeContext } from "next-yak/context/baseContext";

export const useTheme = cache(() => getYakThemeContext());
export const YakThemeProvider = ({ children }: { children: ReactNode }) => {
  return (
    <YakThemeClientProvider theme={useTheme()}>
      {children}
    </YakThemeClientProvider>
  );
};
