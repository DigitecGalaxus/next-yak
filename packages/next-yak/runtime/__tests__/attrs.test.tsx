import React, { FunctionComponent } from "react";
import TestRenderer from "react-test-renderer";
import { beforeEach, expect, it, vi } from "vitest";
import { YakThemeProvider } from "../context";
import { styled } from "../styled";

beforeEach(() => {
  vi.spyOn(console, "warn");
});

type DataAttributes = { [key: `data-${string}`]: any };

it("works fine with an empty object", () => {
  const Comp = styled.div.attrs({})``;
  expect(TestRenderer.create(<Comp />).toJSON()).toMatchInlineSnapshot(`
    <div
      style={{}}
    />
  `);
});

it("works fine with a function that returns an empty object", () => {
  const Comp = styled.div.attrs(() => ({}))``;
  expect(TestRenderer.create(<Comp />).toJSON()).toMatchInlineSnapshot(`
    <div
      style={{}}
    />
  `);
});

it("pass a simple attr via object", () => {
  const Comp = styled.button.attrs({
    type: "button",
  })``;
  expect(TestRenderer.create(<Comp />).toJSON()).toMatchInlineSnapshot(`
    <button
      style={{}}
      type="button"
    />
  `);
});

it("pass a simple attr via function with object return", () => {
  const Comp = styled.button.attrs(() => ({
    type: "button",
  }))``;
  expect(TestRenderer.create(<Comp />).toJSON()).toMatchInlineSnapshot(`
    <button
      style={{}}
      type="button"
    />
  `);
});

it("pass a React component", () => {
  const ReactComponent: FunctionComponent = () => <p>React Component</p>;

  type ButtonProps = {
    component: FunctionComponent;
  };

  const Button = ({ component: ChildComponent }: ButtonProps) => (
    <button>
      <ChildComponent />
    </button>
  );

  const Comp = styled(Button).attrs<Partial<ButtonProps>>(() => ({
    component: ReactComponent,
  }))``;

  expect(TestRenderer.create(<Comp />).toJSON()).toMatchInlineSnapshot(`
    <button>
      <p>
        React Component
      </p>
    </button>
  `);
});

it("should not call a function passed to attrs as an object value", () => {
  const stub = vi.fn(() => "div");

  const Comp = styled.button.attrs<{ foo?: typeof stub }>(() => ({
    foo: stub,
  }))``;

  TestRenderer.create(<Comp />);

  expect(stub).not.toHaveBeenCalled();
});

it("defaultProps are merged into what function attrs receives", () => {
  const Comp = styled.button.attrs<DataAttributes>((props) => ({
    "data-color": props.color,
  }))``;

  Comp.defaultProps = {
    color: "red",
  };

  expect(TestRenderer.create(<Comp />).toJSON()).toMatchInlineSnapshot(`
    <button
      color="red"
      data-color="red"
      style={{}}
    />
  `);
});

it("pass props to the attr function", () => {
  const Comp = styled.button.attrs<{ $submit?: boolean }>((p) => ({
    type: p.$submit ? "submit" : "button",
  }))``;

  expect(TestRenderer.create(<Comp />).toJSON()).toMatchInlineSnapshot(`
    <button
      style={{}}
      type="button"
    />
  `);
  expect(TestRenderer.create(<Comp $submit />).toJSON()).toMatchInlineSnapshot(
    `
    <button
      style={{}}
      type="submit"
    />
  `,
  );
});

it("should replace props with attrs", () => {
  const Comp = styled.button.attrs<{ $submit?: boolean }>((p) => ({
    type: p.$submit ? "submit" : "button",
    tabIndex: 0,
  }))``;

  expect(TestRenderer.create(<Comp />).toJSON()).toMatchInlineSnapshot(`
    <button
      style={{}}
      tabIndex={0}
      type="button"
    />
  `);
  expect(TestRenderer.create(<Comp type="reset" />).toJSON())
    .toMatchInlineSnapshot(`
      <button
        style={{}}
        tabIndex={0}
        type="button"
      />
    `);
  expect(TestRenderer.create(<Comp type="reset" tabIndex={-1} />).toJSON())
    .toMatchInlineSnapshot(`
      <button
        style={{}}
        tabIndex={0}
        type="button"
      />
    `);
});

it("should merge className", () => {
  const Comp = styled.div.attrs(() => ({
    className: "meow nya",
  }))``;

  expect(TestRenderer.create(<Comp />).toJSON()).toMatchInlineSnapshot(`
    <div
      className="meow nya"
      style={{}}
    />
  `);
});

it("should merge className from folded attrs", () => {
  const Inner = styled.div.attrs({ className: "foo" })``;

  const Comp = styled(Inner).attrs(() => ({
    className: "meow nya",
  }))``;

  expect(TestRenderer.create(<Comp className="something" />).toJSON())
    .toMatchInlineSnapshot(`
      <div
        className="something foo meow nya"
        style={{}}
      />
    `);
});

it("should merge className even if its a function", () => {
  const Comp = styled.div.attrs<{ $purr?: boolean }>((p) => ({
    className: `meow ${p.$purr ? "purr" : "nya"}`,
  }))``;

  expect(TestRenderer.create(<Comp />).toJSON()).toMatchInlineSnapshot(`
    <div
      className="meow nya"
      style={{}}
    />
  `);
  expect(TestRenderer.create(<Comp $purr />).toJSON()).toMatchInlineSnapshot(`
    <div
      className="meow purr"
      style={{}}
    />
  `);
});

it("should merge style", () => {
  const Comp = styled.div.attrs(() => ({
    style: { color: "red", background: "blue" },
  }))``;

  expect(
    TestRenderer.create(
      <Comp style={{ color: "green", borderStyle: "dotted" }} />,
    ).toJSON(),
  ).toMatchInlineSnapshot(`
    <div
      style={
        {
          "background": "blue",
          "borderStyle": "dotted",
          "color": "red",
        }
      }
    />
  `);
});

it("should work with data and aria attributes", () => {
  const Comp = styled.div.attrs<DataAttributes>(() => ({
    "data-foo": "bar",
    "aria-label": "A simple FooBar",
  }))``;
  expect(TestRenderer.create(<Comp />).toJSON()).toMatchInlineSnapshot(`
    <div
      aria-label="A simple FooBar"
      data-foo="bar"
      style={{}}
    />
  `);
});

it("merge attrs when inheriting SC", () => {
  let i = 0;
  const Parent = styled.button.attrs((p) => {
    expect(i).toEqual(0);
    i++;
    return {
      type: "button",
      tabIndex: 0,
    };
  })``;
  const Child = styled(Parent).attrs((p) => {
    expect(i).toEqual(1);
    i++;
    return {
      type: "submit",
    };
  })``;
  expect(TestRenderer.create(<Child />).toJSON()).toMatchInlineSnapshot(`
    <button
      style={{}}
      tabIndex={0}
      type="submit"
    />
  `);
});

it("pass attrs to style block", () => {
  /* Would be a React Router Link in real life */
  const Comp = styled.a.attrs<DataAttributes>(() => ({
    href: "#",
    "data-active-class-name": "--is-active",
    // @ts-expect-error
  }))("c1", {
    style: {
      "--testVar": (props: any) => props["data-active-class-name"] && "c2",
    },
  });
  expect(TestRenderer.create(<Comp />).toJSON()).toMatchInlineSnapshot(`
    <a
      className="c1"
      data-active-class-name="--is-active"
      href="#"
      style={
        {
          "--testVar": "c2",
        }
      }
    />
  `);
});

it("should pass through children as a normal prop", () => {
  const Comp = styled.div.attrs(() => ({
    children: "Probably a bad idea",
  }))``;
  expect(TestRenderer.create(<Comp />).toJSON()).toMatchInlineSnapshot(`
    <div
      style={{}}
    >
      Probably a bad idea
    </div>
  `);
});

it("should pass through complex children as well", () => {
  const Comp = styled.div.attrs(() => ({
    children: <span>Probably a bad idea</span>,
  }))``;
  expect(TestRenderer.create(<Comp />).toJSON()).toMatchInlineSnapshot(`
    <div
      style={{}}
    >
      <span>
        Probably a bad idea
      </span>
    </div>
  `);
});

it("should override children", () => {
  const X = styled.div``;
  type XProps = React.ComponentProps<typeof X>["children"] & JSX.Element;
  const Comp = styled.div.attrs(() => ({
    children: <span>Amazing</span>,
  }))``;
  expect(TestRenderer.create(<Comp>Something else</Comp>).toJSON())
    .toMatchInlineSnapshot(`
      <div
        style={{}}
      >
        <span>
          Amazing
        </span>
      </div>
    `);
});

it('should shallow merge "style" prop + attr instead of overwriting', () => {
  const Paragraph = styled.p.attrs<{ $fontScale?: number }>((p) => ({
    style: {
      ...p.style,
      fontSize: `${p.$fontScale}em`,
    },
  }))<{ $fontScale: number }>``;

  const Text: FunctionComponent<
    Partial<React.ComponentProps<typeof Paragraph>>
  > = (props) => {
    const fontScale = 4;

    return (
      <Paragraph $fontScale={fontScale} {...props}>
        {props.children}
      </Paragraph>
    );
  };

  const BlueText = styled(Text).attrs(() => ({
    style: {
      color: "blue",
    },
  }))``;

  const rendered = TestRenderer.create(<BlueText>Hello</BlueText>);

  expect(rendered.toJSON()).toMatchInlineSnapshot(`
    <p
      style={
        {
          "color": "blue",
          "fontSize": "4em",
        }
      }
    >
      Hello
    </p>
  `);
});

it("does not pass transient props to HTML element", () => {
  type CompProps = { $textColor: string };

  const Comp = styled.div<CompProps>`
    color: ${(props) => props.$textColor};
  `;

  const StyledComp = styled(Comp).attrs<Partial<CompProps>>(() => ({
    $textColor: "red",
  }))``;

  expect(TestRenderer.create(<StyledComp />).toJSON()).toMatchInlineSnapshot(`
    <div
      style={{}}
    />
  `);
});

it.skip('should apply given "as" prop to the progressive type', () => {
  const Comp = styled.div.attrs<{ as?: any }>({ as: "video" as const })``;

  //@ts-expect-error
  expect(TestRenderer.create(<Comp loop />).toJSON()).toMatchInlineSnapshot(`
    <video
      className=""
      loop={true}
      style={{}}
    />
  `);
});

// our own tests
it("should remap props", () => {
  const Comp = styled.button.attrs<{ primary?: boolean; $submit?: boolean }>(
    (p) => ({
      type: p.$submit ? "submit" : "button",
      $primary: p.primary,
    }),
  )<{ $primary?: boolean }>``;

  expect(TestRenderer.create(<Comp />).toJSON()).toMatchInlineSnapshot(`
    <button
      style={{}}
      type="button"
    />
  `);
  expect(TestRenderer.create(<Comp primary />).toJSON()).toMatchInlineSnapshot(`
    <button
      primary={true}
      style={{}}
      type="button"
    />
  `);

  expect(TestRenderer.create(<Comp $submit />).toJSON()).toMatchInlineSnapshot(`
    <button
      style={{}}
      type="submit"
    />
  `);
});

it("should have optional attrs props as component interface", () => {
  const Comp = styled.h1.attrs<{ $primary?: boolean }>({
    $primary: true,
  })``;

  expect(TestRenderer.create(<Comp />).toJSON()).toMatchInlineSnapshot(`
    <h1
      style={{}}
    />
  `);

  expect(TestRenderer.create(<Comp $primary />).toJSON())
    .toMatchInlineSnapshot(`
      <h1
        style={{}}
      />
    `);
});

it("should have access to theme", () => {
  const ThemePrinter = ({ theme, ...props }: { theme?: unknown }) => (
    <pre {...props}>{JSON.stringify(theme)}</pre>
  );
  const Comp = styled(ThemePrinter).attrs<DataAttributes>((p) => ({
    "data-color": (p.theme as { color: string }).color,
  }))``;

  expect(
    TestRenderer.create(
      <YakThemeProvider theme={{ color: "red" }}>
        <Comp />
      </YakThemeProvider>,
    ).toJSON(),
  ).toMatchInlineSnapshot(`
    <pre
      data-color="red"
      style={{}}
    />
  `);
});

it("should pass theme if theme is overwritten", () => {
  const ThemePrinter = ({ theme, ...props }: { theme?: unknown }) => (
    <pre {...props}>{JSON.stringify(theme)}</pre>
  );
  const Comp = styled(ThemePrinter).attrs({
    theme: { color: "blue" },
  })``;

  expect(
    TestRenderer.create(
      <YakThemeProvider theme={{ color: "red" }}>
        <Comp />
      </YakThemeProvider>,
    ).toJSON(),
  ).toMatchInlineSnapshot(`
    <pre
      style={{}}
    >
      {"color":"blue"}
    </pre>
  `);
});

it("should allow delete a prop", () => {
  const Comp = styled.h1.attrs<{ primary?: boolean }>({
    primary: undefined,
  })``;

  expect(TestRenderer.create(<Comp primary />).toJSON()).toMatchInlineSnapshot(`
    <h1
      style={{}}
    />
  `);
});

it("should allow rename a prop", () => {
  const Comp = styled.h1.attrs<{ primary?: boolean }>((p) => ({
    ...p,
    primary: undefined,
    $primary: p.primary,
  }))``;

  expect(TestRenderer.create(<Comp primary />).toJSON()).toMatchInlineSnapshot(`
    <h1
      style={{}}
    />
  `);
});

it("should allow rename a prop", () => {
  const Child = styled.h1.attrs<{ primary?: boolean }>((p) => {
    // expect(p.primary).toBe(1);
    return {
      ...p,
      primary: undefined,
      $primary: p.primary,
    };
  })``;
  const Comp = styled(Child).attrs<{ primary?: boolean }>((p) => {
    // expect(p.primary).toBe(2);
    return {};
  })``;

  expect(TestRenderer.create(<Comp primary />).toJSON()).toMatchInlineSnapshot(`
    <h1
      style={{}}
    />
  `);
});
