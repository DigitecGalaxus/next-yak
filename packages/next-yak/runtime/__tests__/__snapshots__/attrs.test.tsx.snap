// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`pass attrs to style block 1`] = `
<a
  className=""
  data-active-class-name="--is-active"
  href="#"
  style={{}}
/>
`;
